# Thông tin đồ án

## Thành viên

|   MSSV   |      Họ tên     |   GitHub  |
|:--------:|:---------------:|:---------:|
| 18120120 | Trần Chí Hào    | haotrch   |
| 18120147 | Trần Quốc Thắng | tqthangbl |

## Link thùng chứa GitHub

[https://github.com/tqthangbl/P4DS-FinalProject](https://github.com/tqthangbl/P4DS-FinalProject)

# Môi trường code



- Code trên môi trường Linux
- Dùng phiên bản các package như trong file "min_ds-env.yml"


In [1]:
import sys
sys.executable

'/home/family-u/.conda/envs/min_ds-env/bin/python'

# Import các thư viện cần thiết

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Thu thập dữ liệu

**Dữ liệu của bạn là về chủ đề gì và bạn lấy từ nguồn nào?**

Bộ dữ liệu mà nhóm sử dụng có tên là "Google Play Store Apps", được lấy từ [Kaggle](https://www.kaggle.com/lava18/google-play-store-apps), chứa thông tin về hơn 10000 ứng dụng trên Google Play Store. Thời điểm bộ dữ liệu được đăng lên Kaggle là 2018-09-05, lần cập nhật cuối là vào 2019-02-03.

Có 2 file dữ liệu:
1. `googleplaystore.csv`: Chứa các thông tin chi tiết về các ứng dụng thu thập được từ Google Play
2. `googleplaystore_user_reviews`: Chứa thông tin về 100 đánh giá liên quan nhất cho mỗi ứng dụng thu thập được từ Google Play

**Người ta có cho phép bạn dùng dữ liệu như này không? Ví dụ, bạn có thể xem thử license của dữ liệu là gì**


License của dữ liệu là Creative Commons Attribution 3.0 Unported, theo như được nói ở [đây](https://creativecommons.org/licenses/by/3.0/), người ta cho phép:

1. Copy và chia sẻ dữ liệu này dưới bất kì hình thức nào
2. Sử dụng bộ dữ liệu này để xây dựng bất cứ thứ gì, cho bất cứ mục đích nào, kể cả thương mại

**Người ta thu thập dữ liệu như thế nào?**

Dữ liệu được tác giả crawl từ Google Play Store.


# Khám phá dữ liệu (thường đan xen với tiền xử lý)

* Dữ liệu có bao nhiêu dòng và bao nhiêu cột?
* Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa
khác nhau không?
* Dữ liệu có các dòng bị lặp không?
* Mỗi cột có ý nghĩa gì?
* Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu
dữ liệu chưa phù hợp để có thể xử lý tiếp không?
* Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá
trị được phân bố như thế nào?
    * Số-lượng/tỉ-lệ các giá trị thiếu?
    * Min? max? → Có gì bất thường không?
* Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical),
các giá trị được phân bố như thế nào?
    * Số-lượng/tỉ-lệ các giá trị thiếu?
    * Số lượng các giá trị khác nhau? Show một vài giá trị
→ Có gì bất thường không?

## Đọc dữ liệu từ file csv vào dataframe

Ở đây nhóm khám phá dữ liệu trong file `googleplaystore.csv` trước. Dữ liệu đọc từ file sẽ lưu vào dataframe `app_df`.

In [86]:
app_df = pd.read_csv('Data/googleplaystore.csv')
app_df.head()

App               Photo Editor & Candy Camera & Grid & ScrapBook
Category                                          ART_AND_DESIGN
Rating                                                       4.1
Reviews                                                      159
Size                                                         19M
Installs                                                 10,000+
Type                                                        Free
Price                                                          0
Content Rating                                          Everyone
Genres                                              Art & Design
Last Updated                                     January 7, 2018
Current Ver                                                1.0.0
Android Ver                                         4.0.3 and up
Name: 0, dtype: object

## Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [4]:
num_rows = len(app_df)
num_cols = len(app_df.columns)

print(f'[+] num_rows = {num_rows}')
print(f'[+] num_cols = {num_cols}')

[+] num_rows = 10841
[+] num_cols = 13


Dữ liệu có 10841 dòng và 13 cột.

## Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

In [5]:
app_df.sample(5)

App        Category  Rating Reviews  \
997                    Funny Quotes Free   ENTERTAINMENT     4.2   23666   
1466                  Redfin Real Estate  HOUSE_AND_HOME     4.6   36857   
7383                      Thistletown CI    PRODUCTIVITY     1.0       1   
5363                       I Am Innocent            GAME     4.4  208501   
4826  Zombie War Z : Hero Survival Rules            GAME     3.9    1987   

      Size    Installs  Type Price Content Rating         Genres  \
997   3.6M  1,000,000+  Free     0           Teen  Entertainment   
1466   19M  1,000,000+  Free     0       Everyone   House & Home   
7383  6.6M        100+  Free     0       Everyone   Productivity   
5363   56M  1,000,000+  Free     0           Teen      Adventure   
4826   62M    100,000+  Free     0     Mature 17+         Action   

           Last Updated Current Ver   Android Ver  
997         May 4, 2018         8.1    4.0 and up  
1466      July 25, 2018       220.0    5.0 and up  
7383     March 15, 2018        41.9    4.1 and up  
5363      July 23, 2018     2.8.307  4.0.3 and up  
4826  December 13, 2017         1.8  4.0.3 and up

Theo như mô tả của tác giả, mỗi dòng trong dữ liệu chứa thông tin của 1 ứng dụng trên Play Store. Bằng việc quan sát ngẫu nhiên các dòng trong dữ liệu nhiều lần, có vẻ không có dòng nào mang ý nghĩa khác. 

## Mỗi cột có ý nghĩa gì?

Theo như mô tả trên Kaggle, bộ dữ liệu gồm 13 cột với ý nghĩa từng cột như sau:

- **App**: Application name
- **Category**: Category the app belongs to 
- **Rating**: Overall user rating of the app (as when scraped)
- **Reviews**: Number of user reviews for the app (as when scraped)
- **Size**: Size of the app (as when scraped)
- **Installs**: Number of user downloads/installs for the app (as when scraped)
- **Type**: Paid or Free
- **Price**: Price of the app (as when scraped)
- **Content Rating**: Age group the app is targeted at - Children / Mature 21+ / Adult
- **Genres**: An app can belong to multiple genres (apart from its main category). For eg, a musical family game will belong to Music, Game, Family genres.

3 cột còn lại không có mô tả trên Kaggle, nhưng dựa vào tên cột và quan sát các dòng ngẫu nhiên, có thể hiểu ý nghĩa của chúng như sau:

- **Last Updated**: Lần cuối ứng dụng được cập nhật trên Play Store
- **Current Ver**: Phiên bản hiện tại của ứng dụng trên Play Store
- **Android Ver**: Các phiên bản Android phù hợp để cài đặt ứng dụng

## Dữ liệu có các dòng bị lặp không?

In [10]:
have_duplicated_rows = True
dup_df = app_df[app_df.duplicated()]
if len(dup_df) == 0:
    have_duplicated_rows = False
    
print(f'[+] have_duplicated_rows = {have_duplicated_rows}')

[+] have_duplicated_rows = True


### Vấn đề cần tiền xử lý
- Một số dòng trong dữ liệu bị lặp, điều này có thể dẫn đến khi tính toán (đếm, tính tỉ lệ). Tuy là Play Store cho phép 2 ứng dụng có thể trùng tên, nhưng rất khó để 2 ứng dụng có thể trùng toàn bộ thông tin trong 12 cột còn lại. Do đó những dòng bị lặp này là của cùng 1 ứng dụng, vì vậy cần phải loại bỏ chúng.

---

# Tiền xử lý

## Loại bỏ các dòng bị lặp

Các dòng bị lặp được lưu trong dataframe `dup_df`. Đầu tiên xem thử 1 vài giá trị của dataframe này.

In [17]:
dup_df.sample(5)

App            Category  \
3811                     CNN Breaking US & World News  NEWS_AND_MAGAZINES   
2645                                            LOVOO              SOCIAL   
2778                      JackThreads: Men's Shopping            SHOPPING   
2745  Shopkick: Free Gift Cards, Shop Rewards & Deals            SHOPPING   
474                       LINE: Free Calls & Messages       COMMUNICATION   

      Rating   Reviews                Size      Installs  Type Price  \
3811     4.0    293080                 25M   10,000,000+  Free     0   
2645     4.0    852455                 68M   10,000,000+  Free     0   
2778     4.0     13085                2.7M    1,000,000+  Free     0   
2745     4.3    213735                 43M   10,000,000+  Free     0   
474      4.2  10790289  Varies with device  500,000,000+  Free     0   

     Content Rating            Genres      Last Updated         Current Ver  \
3811   Everyone 10+  News & Magazines    August 6, 2018                5.17   
2645     Mature 17+            Social    August 3, 2018                27.0   
2778           Teen          Shopping  November 8, 2017                 5.1   
2745       Everyone          Shopping     July 24, 2018              5.3.73   
474        Everyone     Communication     July 26, 2018  Varies with device   

             Android Ver  
3811          4.4 and up  
2645          4.2 and up  
2778        4.0.3 and up  
2745          4.4 and up  
474   Varies with device

Tiến hành loại bỏ các dòng bị lặp trong `app_df` và cập nhật lại biến `num_rows` thể hiện số dòng của dữ liệu.

In [21]:
app_df = app_df.drop_duplicates()
num_rows = len(app_df)

print(f'[+] num_rows = {num_rows}')

[+] num_rows = 10358


Số dòng của dữ liệu sau khi loại bỏ các dòng bị lặp, chỉ còn lại 10358 dòng.

---

# Khám phá dữ liệu (tiếp tục)

## Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

Trước hết lấy ra `dtypes` của các cột, lưu vào series `col_dtypes`, và in ra để quan sát.

In [25]:
col_dtypes = app_df.dtypes

col_dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

Cột `Rating` thể hiện đánh giá của người dùng cho ứng dụng, là 1 con số nên kiểu `float64` là hợp lý. Các cột còn lại đang có kiểu `object`. Kiểu `object` trong Pandas thường là chuỗi, nhưng cũng có thể là một đối tượng bất kỳ. Do đó, trong các cột thuộc kiểu `object`, có thể chứa những giá trị thuộc các kiểu dữ liệu khác nhau. Ta phải kiểm tra điều này bằng cách: với mỗi cột, áp dụng lệnh `type` lên từng phần tử trong cột đó để xem kiểu dữ liệu thật sự của chúng. Hàm `open_object_dtype` sẽ nhận input là 1 series `s`, trả về 1 set là tập các kiểu dữ liệu có trong series đó. Khi xem kết quả hàm `open_object_dtype` của từng cột, ta cũng xem luôn cột đó mang các giá trị phân biệt nào để xem có điều gì bất thường hay không.

In [46]:
def open_object_dtype(s):
    return set(s.apply(type).unique())

for col_name in col_dtypes.index:
    if col_dtypes[col_name] == 'object':
        print(f'[+] Column {col_name} dtypes: ')
        print(open_object_dtype(app_df[col_name]))
        print(app_df[col_name].unique())
        print()

[+] Column App dtypes: 
{<class 'str'>}
['Photo Editor & Candy Camera & Grid & ScrapBook' 'Coloring book moana'
 'U Launcher Lite – FREE Live Cool Themes, Hide Apps' ...
 'Parkinson Exercices FR' 'The SCP Foundation DB fr nn5n'
 'iHoroscope - 2018 Daily Horoscope & Astrology']

[+] Column Category dtypes: 
{<class 'str'>}
['ART_AND_DESIGN' 'AUTO_AND_VEHICLES' 'BEAUTY' 'BOOKS_AND_REFERENCE'
 'BUSINESS' 'COMICS' 'COMMUNICATION' 'DATING' 'EDUCATION' 'ENTERTAINMENT'
 'EVENTS' 'FINANCE' 'FOOD_AND_DRINK' 'HEALTH_AND_FITNESS' 'HOUSE_AND_HOME'
 'LIBRARIES_AND_DEMO' 'LIFESTYLE' 'GAME' 'FAMILY' 'MEDICAL' 'SOCIAL'
 'SHOPPING' 'PHOTOGRAPHY' 'SPORTS' 'TRAVEL_AND_LOCAL' 'TOOLS'
 'PERSONALIZATION' 'PRODUCTIVITY' 'PARENTING' 'WEATHER' 'VIDEO_PLAYERS'
 'NEWS_AND_MAGAZINES' 'MAPS_AND_NAVIGATION' '1.9']

[+] Column Reviews dtypes: 
{<class 'str'>}
['159' '967' '87510' ... '603' '1195' '398307']

[+] Column Size dtypes: 
{<class 'str'>}
['19M' '14M' '8.7M' '25M' '2.8M' '5.6M' '29M' '33M' '3.1M' '28M' '12M

### Vấn đề cần tiền xử lý

Quan sát kết quả, khoan nói đến kiểu dữ liệu ở từng cột, ta thấy có một số điểm bất thường khi xem các giá trị phân biệt ở từng cột:
- Ở cột `Category`, có giá trị là `1.9`, có vẻ không phải là 1 giá trị phù hợp với ý nghĩa category
- Ở cột `Size`, có giá trị là `1,000+`, so với các giá trị còn lại trong cột này thường có format là 1 con số, theo sau đó là đơn vị (`k` hoặc `M`). Giá trị `1,000+` này có vẻ phù hợp với format của cột ngay sau đó là `Install` hơn.
- Ở cột `Installs`, có 1 giá trị là `Free`, không nói lên số lượt cài đặt của ứng dụng, giá trị này lại phù hợp với cột tiếp theo là `Type`, mang ý nghĩa ứng dụng đó thuộc loại miễn phí hay trả phí thì đúng hơn.

Từ đó ta đoán rằng có một dòng nào đó trong dữ liệu bị thiếu giá trị 1 số cột, do đó các giá trị của các cột còn lại bị dịch qua trái, nên giá trị ở một cột này lại nằm ở cột ngay trước đó. Ta cần tiền xử lý để tìm và sửa các dòng như vậy.

# Tiền xử lý

## Tìm các dòng có giá trị một số cột bị thiếu

Ta sử dụng thư viện `csv` để đọc file `googleplaystore.csv`, dòng đầu của file sẽ là tên các cột. Với mỗi dòng tiếp theo, nếu số cột trong dòng đó khác với số lượng cột thì đưa chỉ số của dòng vào list `bad_index`. Sau khi chạy code, `bad_index` sẽ chứa các chỉ số dòng có các cột bị thiếu.

In [92]:
import csv

csv_file = open('Data/googleplaystore.csv')
csv_reader = csv.reader(csv_file, delimiter=',')
row_id = -1
col_count = 0
bad_index = []

for row in csv_reader:
    if row_id == -1:
        col_count = len(row)
        print(f"CSV file has {col_count} columns")
    else:
        if len(row) != col_count:
            print(f"Row id = {row_id} has {len(row)} columns")
            bad_index.append(row_id)
    row_id += 1

CSV file has 13 columns
Row id = 10472 has 12 columns


## Thay thế các giá trị cột bị thiếu thành `NaN`

Có 1 dòng bị thiếu giá trị cột, là dòng có index 10472. Do dataframe `app_df` đã bị loại bỏ các dòng trùng, nên thứ tự các dòng trong file này không giống thứ tự trong file gốc. Vì vậy cần load lại `app_df`, và sửa đổi dòng bị thiếu. Sau đó mới loại bỏ lại các dòng bị trùng.

In [125]:
app_df = pd.read_csv('Data/googleplaystore.csv')

app_df.iloc[bad_index]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

Ở dòng 10472, giá trị cột `Category` bị thiếu, nên ta dịch các giá trị bắt đầu từ cột `Category` sang bên phải rồi điền giá trị cho cột `Category` là `NaN`

In [126]:
cols_list = app_df.columns.tolist()
i = len(cols_list) - 1
while i > 1:
    app_df.loc[10472, cols_list[i]] = app_df.loc[10472, cols_list[i - 1]]
    i -= 1
app_df.loc[10472, 'Category'] = np.nan

app_df.iloc[10472]

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                              NaN
Rating                                                1.9
Reviews                                              19.0
Size                                                 3.0M
Installs                                           1,000+
Type                                                 Free
Price                                                   0
Content Rating                                   Everyone
Genres                                                NaN
Last Updated                            February 11, 2018
Current Ver                                        1.0.19
Android Ver                                    4.0 and up
Name: 10472, dtype: object

Dữ liệu đã được sửa xong, bây giờ loại bỏ dòng bị lặp.

# Đưa ra các câu hỏi có ý nghĩa cần trả lời

* Câu hỏi là gì?
* Nếu trả lời được câu hỏi thì sẽ có lợi ích gì?

# Tiền xử lý + phân tích dữ liệu để trả lời cho từng câu hỏi

IndexError: single positional indexer is out-of-bounds

# Nhìn lại quá trình làm đồ án

* Mỗi thành viên: Đã gặp những khó khăn gì? (Hay mọi
chuyện đều thuận lợi)
* Mỗi thành viên: Có học được gì hữu ích? (Hay không
học được gì)
* Nhóm: Nếu có thêm thời gian thì sẽ làm gì?

# Tài liệu tham khảo